File: yukino/skills/catalog.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package skills

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/hangtiancheng/yukino-code/yukino/logger"
	"github.com/hangtiancheng/yukino-code/yukino/utils"
)

var log = logger.CreateChildLogger("skills")

// Catalog is the in-memory registry of all loaded skills. Phase-1 entries
// contain only frontmatter (PromptBody empty, BodyLoaded false); GetFull
// triggers a phase-2 read of the body on each call (hot reload).
type Catalog struct {
	skills  map[string]*Skill
	sources map[string]string // skill name → "builtin" | "user" | "project" | absolute path
	// order preserves load order (user directory first, then project, each in
	// readdir order) so BuildSkillSection emits the same sequence as TS,
	// whose Map iteration is insertion order.
	order       []string
	workDir     string // remembered so Reload can re-scan the same three tiers
	hasReload   bool
	dirModTimes map[string]time.Time // skill directory path → last known modtime
}

func NewCatalog() *Catalog {
	return &Catalog{
		skills:      make(map[string]*Skill),
		sources:     make(map[string]string),
		dirModTimes: make(map[string]time.Time),
	}
}

// Register adds (or overwrites) a skill in the catalog. The source label is
// surfaced by /skills.
func (c *Catalog) Register(s *Skill, source string) {
	if _, exists := c.skills[s.Meta.Name]; !exists {
		c.order = append(c.order, s.Meta.Name)
	}
	c.skills[s.Meta.Name] = s
	c.sources[s.Meta.Name] = source
}

// Get returns the phase-1 skill (frontmatter only). PromptBody may be empty
// if the catalog was loaded in phase-1 mode.
func (c *Catalog) Get(name string) *Skill {
	return c.skills[name]
}

// Has reports whether a skill is loaded (TS: SkillCatalog.has,
// catalog.ts:230-232).
func (c *Catalog) Has(name string) bool {
	_, ok := c.skills[name]
	return ok
}

// GetFull returns the skill with its body loaded. Disk-backed skills hot
// reload exactly like TS catalog.get: when the file's mtime changed since the
// last load, the whole file is re-parsed and BOTH meta and body are refreshed
// (a frontmatter rename triggers a full catalog reload); on read/parse
// failure the cached version is retained. Embedded builtins are cache hits.
func (c *Catalog) GetFull(name string) (*Skill, error) {
	skill, ok := c.skills[name]
	if !ok {
		return nil, fmt.Errorf("unknown skill: %s", name)
	}
	if skill.SourceFile == "" {
		// Embedded skill — body was loaded at startup, nothing to refresh.
		return skill, nil
	}

	// Hot reload: re-read only when the file changed since the last load
	// (TS: entry.loadedMtimeMs > 0 && currentMtime !== entry.loadedMtimeMs;
	// an unreadable mtime skips reloading entirely).
	if skill.LoadedMtimeMs > 0 {
		info, err := os.Stat(skill.SourceFile)
		if err != nil {
			// Retain the cached version if reading fails (TS get()'s catch logs).
			log.Error("skills operation failed", "err", err)
		} else {
			currentMtime := info.ModTime().UnixMilli()
			if currentMtime != skill.LoadedMtimeMs {
				raw, rerr := os.ReadFile(skill.SourceFile)
				if rerr != nil {
					log.Error("skills operation failed", "err", rerr)
				} else if parsed, perr := parseSkillFile(string(raw)); perr == nil {
					if parsed.Meta.Name != name {
						// Rebuild indexes and precedence when frontmatter
						// renames a skill (TS: this.reload()).
						c.Reload(c.workDir)
						renamed, found := c.skills[name]
						if !found {
							return nil, fmt.Errorf("unknown skill: %s", name)
						}
						return renamed, nil
					}
					skill.Meta = parsed.Meta
					skill.PromptBody = parsed.Body
					skill.BodyLoaded = true
					skill.LoadedMtimeMs = currentMtime
					return skill, nil
				}
				// Retain the cached version if parsing fails — a
				// single bad write should not cause a skill to vanish.
			}
		}
	}
	return c.ensureBody(skill)
}

// ensureBody performs the phase-2 body read for entries whose body was never
// loaded, preserving the cached body on failure.
func (c *Catalog) ensureBody(skill *Skill) (*Skill, error) {
	if skill.BodyLoaded {
		return skill, nil
	}
	if err := loadSkillBody(skill); err != nil {
		if skill.PromptBody == "" {
			return nil, err
		}
		return skill, err
	}
	return skill, nil
}

// List returns metadata for every loaded skill in catalog load order
// (registration order, with an overwritten name keeping its first position) —
// exactly like TS's Map insertion order in catalog.list() (catalog.ts:182-184).
func (c *Catalog) List() []SkillMeta {
	result := make([]SkillMeta, 0, len(c.order))
	for _, name := range c.order {
		if s, ok := c.skills[name]; ok {
			result = append(result, s.Meta)
		}
	}
	return result
}

// Source returns the origin label for a skill ("builtin", "user", "project",
// or a path). Returns "" if the skill isn't loaded.
func (c *Catalog) Source(name string) string {
	return c.sources[name]
}

// Reload re-scans all three tiers (builtin + user + project) and rebuilds
// the catalog in place. Used by `/skills reload` and tests.
func (c *Catalog) Reload(workDir string) {
	fresh := LoadCatalog(workDir)
	c.skills = fresh.skills
	c.sources = fresh.sources
	c.order = fresh.order
	c.workDir = fresh.workDir
	c.dirModTimes = fresh.dirModTimes
}

// NeedsReload checks whether the skill directories' modtimes have changed
// since the catalog was last loaded. A changed modtime indicates a skill was
// added or removed (file edits within existing skills are already handled by
// GetFull's per-call re-read). TS needsReload (catalog.ts:77-91): iterate the
// recorded directories; a stat failure reloads only when a real mtime was
// recorded (the zero time stands for TS's null).
func (c *Catalog) NeedsReload() bool {
	for dir, recorded := range c.dirModTimes {
		info, err := os.Stat(dir)
		if err != nil {
			if recorded.IsZero() {
				continue
			}
			return true // directory disappeared
		}
		if !info.ModTime().Equal(recorded) {
			return true
		}
	}
	return false
}

// snapshotDirModTimes records current modtimes of every watched directory:
// the skill-dir roots plus any subdirectory recorded during the scan, so a
// directory created after a missing-dirs load is still detected
// (TS: snapshotDirModTimes, catalog.ts:97-110).
func (c *Catalog) snapshotDirModTimes() {
	watched := make(map[string]bool, len(c.dirModTimes)+2)
	for _, dir := range skillDirPaths(c.workDir) {
		watched[dir] = true
	}
	for dir := range c.dirModTimes {
		watched[dir] = true
	}
	for dir := range watched {
		info, err := os.Stat(dir)
		if err != nil {
			c.dirModTimes[dir] = time.Time{}
			continue
		}
		c.dirModTimes[dir] = info.ModTime()
	}
}

// skillDirPaths returns the user-global and project skill directory paths:
// ~/.agents/skills first, then $workDir/.agents/skills (highest priority)
// (TS: skillDirPaths, catalog.ts:112-116).
func skillDirPaths(workDir string) []string {
	var dirs []string
	if home, err := os.UserHomeDir(); err == nil {
		dirs = append(dirs, filepath.Join(home, ".agents", "skills"))
	}
	if workDir != "" {
		dirs = append(dirs, filepath.Join(workDir, ".agents", "skills"))
	}
	return dirs
}

// LoadCatalog builds a phase-1 catalog by merging three tiers, with later
// sources overriding earlier ones by name (project wins over user wins over
// builtin):
//  1. internal/skills/builtins/* (embedded via go:embed, lowest priority)
//  2. ~/.agents/skills/         (user global)
//  3. $workDir/.agents/skills/  (project, highest priority)
//
// Only frontmatter is read at this stage; PromptBody stays empty until
// GetFull is called. Parse failures on individual skills are silently
// skipped — one bad file must not bring down the whole catalog.
func LoadCatalog(workDir string) *Catalog {
	c := NewCatalog()
	c.workDir = workDir

	// Tier 1: embedded builtins
	for _, s := range LoadBuiltins() {
		c.Register(s, "builtin")
	}

	// Tier 2: user global
	if home, err := os.UserHomeDir(); err == nil {
		loadTierInto(c, filepath.Join(home, ".agents", "skills"), "user")
	}

	// Tier 3: project
	loadTierInto(c, filepath.Join(workDir, ".agents", "skills"), "project")

	c.snapshotDirModTimes()
	return c
}

// LoadFromDirectory loads every subdirectory of dir as a skill. Used by tests
// and one-off callers that just want a single tier loaded. Body is read
// eagerly (no two-phase split) so existing test code that touches
// skill.PromptBody continues to work.
func LoadFromDirectory(dir string) (*Catalog, error) {
	c := NewCatalog()
	loadTierEager(c, dir, dir)
	return c, nil
}

// LoadSkills is the legacy two-tier loader kept for backward compatibility
// with code that still pre-loads bodies eagerly. New callers should use
// LoadCatalog + GetFull. Order: user global → project.
func LoadSkills(workDir string) *Catalog {
	c := NewCatalog()
	c.workDir = workDir
	if home, err := os.UserHomeDir(); err == nil {
		loadTierEager(c, filepath.Join(home, ".agents", "skills"), "user")
	}
	loadTierEager(c, filepath.Join(workDir, ".agents", "skills"), "project")
	return c
}

// loadTierInto walks a single tier and registers each subdir as a phase-1
// skill (TS: scanDirectory + loadSkill, catalog.ts:118-180).
func loadTierInto(c *Catalog, dir, source string) {
	c.scanDirectory(dir, source, false)
}

// loadTierEager is like loadTierInto but also reads the body. Used by legacy
// LoadSkills / LoadFromDirectory to preserve old behavior.
func loadTierEager(c *Catalog, dir, source string) {
	c.scanDirectory(dir, source, true)
}

// scanDirectory mirrors TS scanDirectory (catalog.ts:118-146): every entry is
// stat'ed (following symlinks, like TS statSync); directories are recorded in
// dirModTimes and loaded when they contain a SKILL.md. A broken symlink or a
// concurrently removed entry must not hide other skills.
//
// Residual: os.ReadDir returns entries sorted by name while TS readdirSync
// returns OS order. Later duplicates win in both, so when one tier holds two
// directories declaring the same frontmatter name (or for list()/section
// ordering), the winner/sequence can differ from TS. Reproducing OS order
// would need raw syscalls — deliberately not done.
func (c *Catalog) scanDirectory(dir, source string, eager bool) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if !os.IsNotExist(err) {
			log.Error("skills operation failed", "err", err)
		}
		return
	}
	for _, entry := range entries {
		fullPath := filepath.Join(dir, entry.Name())
		info, err := os.Stat(fullPath)
		if err != nil {
			// A broken symlink or a concurrently removed entry must not hide other skills.
			log.Error("skills operation failed", "err", err)
			continue
		}
		if !info.IsDir() {
			continue
		}
		c.dirModTimes[fullPath] = info.ModTime()
		// TS: existsSync(SKILL.md) — a directory without the manifest is
		// skipped silently.
		if _, err := os.Stat(filepath.Join(fullPath, "SKILL.md")); err != nil {
			continue
		}
		skill, err := parseFrontmatterOnly(fullPath)
		if err != nil {
			continue // read/parse failures are logged where they occur
		}
		if eager {
			_ = loadSkillBody(skill)
		}
		c.Register(skill, source)
	}
}

// xmlEscaper escapes text embedded in the skill XML envelopes
// (TS: escapeSkillXml, catalog.ts:295-300).
var xmlEscaper = strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;")

// EscapeSkillXml escapes &, < and > for the skill XML envelopes.
func EscapeSkillXml(text string) string { return xmlEscaper.Replace(text) }

// BuildSkillSection renders the metadata-only conversation reminder; bodies
// load on demand without changing the system prefix (TS: buildSkillSection,
// catalog.ts:302-329). Returns "" when the catalog is empty.
func BuildSkillSection(catalog *Catalog, workDir string) string {
	if catalog == nil {
		return ""
	}
	metas := catalog.List()
	if len(metas) == 0 {
		return ""
	}
	// Emit in catalog load order (user directory first, then project; each in
	// readdir order), exactly like TS catalog.list() — an alphabetical sort
	// would make the assembled section differ from the TS bytes.

	skillsDir := filepath.Join(workDir, ".agents", "skills")
	lines := []string{
		"## Skills",
		`Load relevant instructions with LoadSkill {name: "<skill-name>"}, or user command /<skill-name>. Mode inline activates in this conversation; fork runs in a subagent when available, otherwise inline. Load resources only as needed, relative to the skill directory. Tool access remains host-controlled.`,
		`InstallSkill {source: "<local path or raw SKILL.md URL>"} makes skills available immediately. skills.sh pages and GitHub tree/blob pages are not supported.`,
		"Create skills under the following directory as <skill-name>/SKILL.md:",
		"<skills-directory>" + EscapeSkillXml(skillsDir) + "</skills-directory>",
		"<available-skills>",
	}
	for _, meta := range metas {
		oneLine := strings.Join(strings.Fields(meta.Description), " ")
		desc := oneLine
		// TS slices by UTF-16 code units (oneLine.length > 200), not runes.
		if utils.UTF16Len(oneLine) > 200 {
			desc = utils.TruncateUTF16(oneLine, 200) + "…"
		}
		mode := meta.Mode
		if mode == "" {
			mode = "inline"
		}
		lines = append(lines,
			"<skill><name>"+EscapeSkillXml(meta.Name)+"</name><description>"+EscapeSkillXml(desc)+"</description><mode>"+mode+"</mode></skill>")
	}
	lines = append(lines, "</available-skills>")
	return strings.Join(lines, "\n")
}

File: apps/yukino/src/skills/catalog.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { readdirSync, readFileSync, existsSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

import yaml from "js-yaml";
import { z } from "zod";

import type { Skill, SkillMeta } from "./index.js";

import { createChildLogger } from "@/logger/index.js";
import { asRecord, strArg } from "@/utils/index.js";

const log = createChildLogger({ module: "skills" });

/**
 * Internal skill storage with source file path and load timestamp for hot reloading
 *
 */
interface CatalogEntry {
  skill: Skill;
  /** Absolute path to SKILL.md, used for re-reading during hot reloading */
  filePath: string;

  /** File modification time (ms) when last loaded. 0 means the mtime could not be read, so hot reloading is skipped */
  loadedMtimeMs: number;
}

export class SkillCatalog {
  private entries = new Map<string, CatalogEntry>();
  private workDir = "";
  private dirModTimes = new Map<string, number | null>();

  load(workDir: string): void {
    this.workDir = workDir;
    this.entries.clear();
    this.dirModTimes.clear();

    // User-global ~/.agents/skills/, then project-level
    // $workDir/.agents/skills/ (highest priority)
    for (const dir of this.skillDirPaths()) {
      if (!existsSync(dir)) {
        continue;
      }

      this.scanDirectory(dir);
    }

    this.snapshotDirModTimes();
  }

  /**
   * Check whether a skill directory mtime has changed (a skill was added or deleted).
   * Edits to existing skill files are handled by lazy re-reading in get().
   */
  needsReload(): boolean {
    for (const [dir, recorded] of this.dirModTimes) {
      try {
        const current = statSync(dir).mtimeMs;
        if (current !== recorded) {
          return true;
        }
      } catch {
        if (recorded !== null) {
          return true;
        }
      }
    }
    return false;
  }

  reload(): void {
    this.load(this.workDir);
  }

  private snapshotDirModTimes(): void {
    // Watch each skill directory too: adding/removing SKILL.md does not change
    // the parent skills directory's mtime.
    for (const dir of new Set([
      ...this.skillDirPaths(),
      ...this.dirModTimes.keys(),
    ])) {
      try {
        this.dirModTimes.set(dir, statSync(dir).mtimeMs);
      } catch {
        this.dirModTimes.set(dir, null);
      }
    }
  }

  private skillDirPaths(): string[] {
    return [homedir(), ...(this.workDir ? [this.workDir] : [])].flatMap(
      (root) => [".agents"].map((ecosystem) => join(root, ecosystem, "skills")),
    );
  }

  private scanDirectory(dir: string) {
    let dirEntries: string[];
    try {
      dirEntries = readdirSync(dir);
    } catch (err) {
      log.error({ err }, "skills operation failed");
      return;
    }

    for (const entry of dirEntries) {
      const fullPath = join(dir, entry);
      try {
        const stat = statSync(fullPath);
        if (stat.isDirectory()) {
          this.dirModTimes.set(fullPath, stat.mtimeMs);
          const skillFile = join(fullPath, "SKILL.md");
          if (existsSync(skillFile)) {
            this.loadSkill(skillFile, fullPath, true);
          }
        }
      } catch (err) {
        // A broken symlink or a concurrently removed entry must not hide other skills.
        log.error({ err }, "skills operation failed");
      }
      // else if (entry.endsWith(".md") && entry !== "SKILL.md") {
      //   this.loadSkill(fullPath, dir, false);
      // }
    }
  }
  private loadSkill(filePath: string, sourceDir: string, isDirectory: boolean) {
    try {
      const raw = readFileSync(filePath, "utf-8");
      const parsed = parseSkillFile(raw);
      if (!parsed) {
        return;
      }

      const skill: Skill = {
        meta: parsed.meta,
        body: parsed.body,
        sourceDir,
        isDirectory,
      };

      // Record file modification time for subsequent hot reload detection
      let mtimeMs = 0;
      try {
        mtimeMs = statSync(filePath).mtimeMs;
      } catch (err) {
        log.error({ err }, "skills operation failed");
        // Fail gracefully if timestamp cannot be retrieved
      }

      this.entries.set(skill.meta.name, {
        skill,
        filePath,
        loadedMtimeMs: mtimeMs,
      });
    } catch (err) {
      log.error({ err }, "skills operation failed");
      // Skip invalid skill
    }
  }

  list(): SkillMeta[] {
    return [...this.entries.values()].map((e) => e.skill.meta);
  }

  /**
   * Gets a skill with hot reload support: automatically re-reads the file if it has been modified on disk.
   * re-reads the body on every call (hot reload),
   * and retains the cached body if reading fails.
   */
  get(name: string): Skill | undefined {
    const entry = this.entries.get(name);
    if (!entry) {
      return undefined;
    }

    // Attempt hot reload: check if the file has been modified
    if (entry.filePath && entry.loadedMtimeMs > 0) {
      try {
        const currentMtime = statSync(entry.filePath).mtimeMs;
        if (currentMtime !== entry.loadedMtimeMs) {
          // File has been modified, re-read it
          const raw = readFileSync(entry.filePath, "utf-8");
          const parsed = parseSkillFile(raw);
          if (parsed) {
            if (parsed.meta.name !== name) {
              // Rebuild indexes and precedence when frontmatter renames a skill.
              this.reload();
              return this.entries.get(name)?.skill;
            }
            entry.skill = {
              meta: parsed.meta,
              body: parsed.body,
              sourceDir: entry.skill.sourceDir,
              isDirectory: entry.skill.isDirectory,
            };
            entry.loadedMtimeMs = currentMtime;
          }
          // Retain the cached version if parsing fails — a single bad write should not cause a skill to vanish
        }
      } catch (err) {
        log.error({ err }, "skills operation failed");
        // Retain the cached version if reading fails
      }
    }

    return entry.skill;
  }

  has(name: string): boolean {
    return this.entries.has(name);
  }
}

/**
 * Normalize the execution mode.
 *
 * Some agent ecosystems use `context: fork` to express "isolated execution", which is
 * semantically equivalent to `mode: fork` here. Both forms are interchangeable, so
 * externally sourced skills work without modification.
 */
function resolveMode(raw: unknown): "inline" | "fork" {
  // raw.mode
  const mode = strArg(asRecord(raw), "mode");
  if (mode === "inline" || mode === "fork") {
    return mode;
  }
  // raw.context
  return strArg(asRecord(raw), "context") === "fork" ? "fork" : "inline";
}

const YamlFrontmatterSchema = z.looseObject({
  name: z.string().trim().min(1),
  description: z.string().optional(),
  mode: z.enum(["inline", "fork"]).optional(),
  model: z.string().optional(),
  fork_context: z.enum(["full", "none", "recent"]).optional(),
});

export function parseSkillFile(content: string): {
  meta: SkillMeta;
  body: string;
  frontmatter: Record<string, unknown>;
} | null {
  // Delimiters occupy their own lines; `---` inside YAML strings is content.
  const normalized = content.replace(/^\uFEFF/, "");
  const match = /^---[ \t]*\r?\n([\s\S]*?)\r?\n---[ \t]*(?:\r?\n|$)/.exec(
    normalized,
  );
  if (!match) {
    return null;
  }
  const body = normalized.slice(match[0].length).trim();

  try {
    const raw: unknown = yaml.load(match[1]);
    const data = YamlFrontmatterSchema.parse(raw);
    return {
      meta: {
        name: data.name,
        description: data.description ?? "",
        mode: resolveMode(raw),
        model: data.model,
        forkContext: data.fork_context,
      },
      body,
      frontmatter: data,
    };
  } catch (err) {
    log.error({ err }, "skills operation failed");
    return null;
  }
}

export function escapeSkillXml(text: string): string {
  return text
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;");
}

/** Metadata-only conversation reminder; bodies load on demand without changing the system prefix. */
export function buildSkillSection(
  catalog: SkillCatalog,
  workDir: string,
): string {
  const metas = catalog.list();
  if (metas.length === 0) {
    return "";
  }
  const skillsDir = join(workDir, ".agents", "skills");
  const lines = [
    "## Skills",
    'Load relevant instructions with LoadSkill {name: "<skill-name>"}, or user command /<skill-name>. Mode inline activates in this conversation; fork runs in a subagent when available, otherwise inline. Load resources only as needed, relative to the skill directory. Tool access remains host-controlled.',
    'InstallSkill {source: "<local path or raw SKILL.md URL>"} makes skills available immediately. skills.sh pages and GitHub tree/blob pages are not supported.',
    "Create skills under the following directory as <skill-name>/SKILL.md:",
    `<skills-directory>${escapeSkillXml(skillsDir)}</skills-directory>`,
    "<available-skills>",
  ];
  for (const meta of metas) {
    const oneLine = meta.description.replace(/\s+/g, " ").trim();
    const desc = oneLine.length > 200 ? oneLine.slice(0, 200) + "…" : oneLine;
    lines.push(
      `<skill><name>${escapeSkillXml(meta.name)}</name><description>${escapeSkillXml(desc)}</description><mode>${meta.mode ?? "inline"}</mode></skill>`,
    );
  }
  lines.push("</available-skills>");
  return lines.join("\n");
}
